eclair-core/src/main/scala/fr/acinq/eclair/Features.scala

```diff
@@ -38,18 +35,24 @@ object Features {
   val VARIABLE_LENGTH_ONION_MANDATORY = 8
   val VARIABLE_LENGTH_ONION_OPTIONAL = 9
 
-  def hasFeature(features: BitSet, bit: Int): Boolean = features.get(bit)
+  // Note that BitVector indexes from left to right whereas the specification indexes from right to left.
+  // This is why we have to reverse the bits to check if a feature is set.
+
+  def hasFeature(features: BitVector, bit: Int): Boolean = if (features.sizeLessThanOrEqual(bit)) false else features.reverse.get(bit)
 
-  def hasFeature(features: ByteVector, bit: Int): Boolean = hasFeature(BitSet.valueOf(features.reverse.toArray), bit)
+  def hasFeature(features: ByteVector, bit: Int): Boolean = hasFeature(features.bits, bit)
+
+  def hasVariableLengthOnion(features: ByteVector): Boolean = hasFeature(features, VARIABLE_LENGTH_ONION_MANDATORY) || hasFeature(features, VARIABLE_LENGTH_ONION_OPTIONAL)
```

Comment:
This isn't consistent with what we currently have in the `Peer`:

https://github.com/ACINQ/eclair/blob/71da6776e133e1c4caf18f2c1ac091b48c4f5c8f/eclair-core/src/main/scala/fr/acinq/eclair/io/Peer.scala#L137-L138

https://github.com/ACINQ/eclair/blob/71da6776e133e1c4caf18f2c1ac091b48c4f5c8f/eclair-core/src/main/scala/fr/acinq/eclair/io/Peer.scala#L154

Not sure what's the best approach.

Reply:
Strictly speaking it would be more correct to handle `variable_length_onion_optional` and `variable_length_onion_mandatory` separately, when mandatory is set we should reject legacy payloads.
I don't think anyone wants to set mandatory at the moment though, I don't know if we should correctly implement the distinction between the two.
But you're right that separating the two checks makes it more explicit that we're not handling the distinction yet, I'll change that.